### DISCUSSION TITLE
Custom Input deserialization

### DISCUSSION THREAD

**Discussion by lukas-krecan** [Netflix/dgs-framework#699] [Ideas]
I am looking for a way how to customize input deserialization. I have 2 use cases

1. Custom IDs - In Kotlin code we are trying to have different types for IDs like `ShowId` and `ActorId`. Currently we can not use them in Input objects as there is no way how to do it. (We can not use custom scalars as we are using Relay node query so the IDs have to be of the same type in the schema). We can use custom primary constructor for that, but it's polluting the classes.
2. Null vs. missing input field - In some cases we would like to differentiate between an absent input (not specified) field and explicitly specified null. (Not sure if it's a good idea, but we want to know if the user wants to clear a nullable field by setting null or if he wants to keep the previous value).

I have checked the source code and all the relevant methods are private and members of a final class so there is basically no way for us to customize it. I am basically some ability to change the behavior of `DataFetcherInvoker.convertValue`

**Comment by lukas-krecan** [comment]
It seems that Custom IDs can be implemented by custom scalar parser that is registers on the ID type. I did not know it's possible to override the default Scalar

**Comment by lukas-krecan** [comment]
For the nullable use-case we can get the value from DgsDataFetchingEnvironment.arguments

**Comment by srinivasankavitha** [comment]
Thanks for the post and your findings. I was not aware of this set up
either, so that's really useful to know it's available. I would also be
interested in looking at your code when you have it set up so we can
document that for future reference.
